client/host_queue.go

```diff
@@ -0,0 +1,265 @@
+			writeBatchSize    = q.opts.GetWriteBatchSize()
+			opsLen            = len(ops)
+		)
+		for i := 0; i < opsLen; i++ {
+			switch v := ops[i].(type) {
+			case *writeOp:
+				if currWriteOps == nil {
+					currWriteOps = q.opsArrayPool.Get()
+					currWriteRequests = q.writeRequestArrayPool.Get()
+				}
+
+				currWriteOps = append(currWriteOps, ops[i])
+				currWriteRequests = append(currWriteRequests, &v.request)
+
+				if len(currWriteOps) == writeBatchSize {
+					// Reached write batch limit, write async and reset
+					q.asyncWrite(wgAll, currWriteOps, currWriteRequests)
+					currWriteOps = nil
+					currWriteRequests = nil
+				}
+			default:
+				completionFn := ops[i].GetCompletionFn()
+				completionFn(nil, errUnknownOperation)
+			}
+		}
+
+		// If any outstanding write ops, async write
+		if len(currWriteOps) > 0 {
+			q.asyncWrite(wgAll, currWriteOps, currWriteRequests)
+		}
+
+		q.opsArrayPool.Put(ops)
+
+		q.RLock()
+		closed := q.closed
+		q.RUnlock()
+
+		if closed {
+			// Final drain, close the connection pool after all requests done
+			wgAll.Wait()
+			q.connPool.Close()
+			return
+		}
+	}
+}
+
+func (q *queue) asyncWrite(wg *sync.WaitGroup, ops []op, elems []*rpc.WriteRequest) {
+	wg.Add(1)
+	// TODO(r): Use a worker pool to avoid creating new go routines for async writes
+	go func() {
+		req := q.writeBatchRequestPool.Get()
+		req.Elements = elems
+
+		// NB(r): Defer is slow in the hot path unfortunately
+		cleanup := func() {
+			q.writeBatchRequestPool.Put(req)
+			q.writeRequestArrayPool.Put(elems)
+			q.opsArrayPool.Put(ops)
+			wg.Done()
+		}
+
+		client, err := q.connPool.NextClient()
+		if err != nil {
+			callAllCompletionFns(ops, nil, err)
+			cleanup()
+			return
+		}
+
+		ctx, _ := thrift.NewContext(q.opts.GetWriteRequestTimeout())
+		if err := client.WriteBatch(ctx, req); err != nil {
+			if batchErrs, ok := err.(*rpc.WriteBatchErrors); ok {
+				// Callback all writes with errors
+				hasErr := make(map[int]struct{})
+				for _, batchErr := range batchErrs.Errors {
+					op := ops[batchErr.ElementErrorIndex]
+					op.GetCompletionFn()(nil, fmt.Errorf(batchErr.Error.Message))
+					hasErr[int(batchErr.ElementErrorIndex)] = struct{}{}
+				}
+				// Callback all writes with no errors
+				for i := range ops {
+					if _, ok := hasErr[i]; !ok {
+						// No error
+						ops[i].GetCompletionFn()(nil, nil)
+					}
+				}
+			} else {
+				// Entire batch failed
+				callAllCompletionFns(ops, nil, err)
+			}
+			cleanup()
+			return
+		}
+
+		// All succeeded
+		callAllCompletionFns(ops, nil, nil)
+		cleanup()
+	}()
+}
+
+func (q *queue) Enqueue(o op) error {
+	q.Lock()
+	if q.closed {
+		q.Unlock()
+		return errQueueClosed
+	}
+	q.ops = append(q.ops, o)
+	// If queue is full flush
+	if len(q.ops) == q.size {
+		q.flushWithLock()
+	}
+	q.Unlock()
+	return nil
+}
+
+func (q *queue) GetConnectionCount() int {
+	return q.connPool.GetConnectionCount()
+}
+
+func (q *queue) Close() {
+	q.Lock()
```

Comment:
check against `q.closed` here perhaps? If `Close` is called twice, we don't wanna be stuck in `flushWithLock` if the `drain` goroutine has already exited.